cfgov/v1/views/reports.py

```diff
@@ -49,3 +60,49 @@ def get_filename(self):
     def get_queryset(self):
         return CFGOVPage.objects.filter(live=True).prefetch_related(
             "tags", "categories")
+
+
+class DocumentsReportView(ReportView):
+    header_icon = "doc-full"
+    title = "All documents"
+
+    list_export = [
+        "id",
+        "title",
+        "filename",
+        "url",
+        "collection",
+        "tags.names",
+        "created_at",
+        "uploaded_by_user",
+    ]
+    export_headings = {
+        "id": "ID",
+        "title": "Title",
+        "filename": "File",
+        "url": "URL",
+        "collection": "Collection",
+        "tags.names": "Tags",
+        "created_at": "Uploaded on",
+        "uploaded_by_user": "Uploaded by",
+    }
+
+    """TODO: Why must tags be processed here but not page metadata report?"""
+    custom_field_preprocess = {
+        "tags.names": {
+            "csv": process_tags
+        },
```

Comment:
There seems to be a difference with how `ReportView` and `PageReportView` handle `tags.names` in the CSV export. As far as I can tell, `PageReportView` spits out a nice comma-separated list for `tags.names`, while `ReportView` returns a queryset of tag names for it. I'm not 100% sure why that's the case—or if I'm misunderstanding something—but I do know using just `tags.names` for the CSV export like the page metadata report does on line 34 above doesn't work here—it puts a stringified queryset in that column of the CSV. To work around that, I add a simple `process_tags` function to preprocess the queryset of tag names into a comma-separated string. But if there's a better way to do that, I'll happily change (and if there's not, I'll remove the TODO comment).

Reply:
This is a fascinating rabbit hole I've gone down — there' appears to be [nothing specific to the `PageReportView` around tags or/related models at all](https://github.com/wagtail/wagtail/blob/main/wagtail/admin/views/reports/base.py#L63-L75). So, I started looking at the mixins, since `ReportView` inherits the `custom_field_preprocess` from `SpreadsheetExportMixin`, but there's nothing there either.

However, then I realized, tags in our pages come from `CFGOVPage`, [which uses `CusterableTagManager` for `tags`](https://github.com/cfpb/consumerfinance.gov/blob/main/cfgov/v1/models/base.py#L67-L68), which comes from django-modelcluster. Wagtail's `Document` [uses django-taggit's `TaggableManager `](https://github.com/wagtail/wagtail/blob/e2b30c1690fb231abbb2e896e5553119db24cf8b/wagtail/documents/models.py#L38).

I suspect this related object manager being different is part of the issue. Both of these have different implementations of `value_from_object`: [taggit's](https://github.com/jazzband/django-taggit/blob/master/taggit/managers.py#L559-L565) vs [modelcluster](https://github.com/wagtail/django-modelcluster/blob/main/modelcluster/contrib/taggit.py#L97-L105). I'm not familiar enough with django-taggit to know quite the impact of these differences right off, but that's my suspicion. It's a long way to say, what you've got here is probably fine — however frustrating the difference between pages and documents.

Reply:
Thanks for going down that rabbit hole, @willbarton. I removed the TODO in ef34cdb.